// @electron-forge/template-vite/dist/ViteTemplate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const template_base_1 = require("@electron-forge/template-base");
const fs_extra_1 = __importDefault(require("fs-extra"));
class ViteTemplate extends template_base_1.BaseTemplate {
    constructor() {
        super(...arguments);
        this.templateDir = node_path_1.default.resolve(__dirname, '..', 'tmpl');
    }
    async initializeTemplate(directory, options) {
        const superTasks = await super.initializeTemplate(directory, options);
        return [
            ...superTasks,
            {
                title: 'Setting up Forge configuration',
                task: async () => {
                    await this.copyTemplateFile(directory, 'forge.config.js');
                },
            },
            {
                title: 'Setting up Vite configuration',
                task: async () => {
                    await this.copyTemplateFile(directory, 'vite.main.config.mjs');
                    await this.copyTemplateFile(directory, 'vite.preload.config.mjs');
                    await this.copyTemplateFile(directory, 'vite.renderer.config.mjs');
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'renderer.js');
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'preload.js');
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'index.js');
                    await this.updateFileByLine(node_path_1.default.resolve(directory, 'src', 'index.js'), (line) => {
                        if (line.includes('mainWindow.loadFile'))
                            return `  if (MAIN_WINDOW_VITE_DEV_SERVER_URL) {
    mainWindow.loadURL(MAIN_WINDOW_VITE_DEV_SERVER_URL);
  } else {
    mainWindow.loadFile(path.join(__dirname, \`../renderer/\${MAIN_WINDOW_VITE_NAME}/index.html\`));
  }`;
                        return line;
                    }, node_path_1.default.resolve(directory, 'src', 'main.js'));
                    // TODO: Compatible with any path entry.
                    // Vite uses index.html under the root path as the entry point.
                    fs_extra_1.default.moveSync(node_path_1.default.join(directory, 'src', 'index.html'), node_path_1.default.join(directory, 'index.html'), { overwrite: options.force });
                    await this.updateFileByLine(node_path_1.default.join(directory, 'index.html'), (line) => {
                        if (line.includes('link rel="stylesheet"'))
                            return '';
                        if (line.includes('</body>'))
                            return '    <script type="module" src="/src/renderer.js"></script>\n  </body>';
                        return line;
                    });
                    // update package.json entry point
                    const pjPath = node_path_1.default.resolve(directory, 'package.json');
                    const currentPJ = await fs_extra_1.default.readJson(pjPath);
                    currentPJ.main = '.vite/build/main.js';
                    await fs_extra_1.default.writeJson(pjPath, currentPJ, {
                        spaces: 2,
                    });
                },
            },
        ];
    }
}
exports.default = new ViteTemplate();
