// @electron-forge/maker-squirrel/dist/MakerSquirrel.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakerSquirrel = void 0;
const node_path_1 = __importDefault(require("node:path"));
const maker_base_1 = require("@electron-forge/maker-base");
const electron_winstaller_1 = require("electron-winstaller");
const fs_extra_1 = __importDefault(require("fs-extra"));
class MakerSquirrel extends maker_base_1.MakerBase {
    constructor() {
        super(...arguments);
        this.name = 'squirrel';
        this.defaultPlatforms = ['win32'];
    }
    isSupportedOnCurrentPlatform() {
        return (this.isInstalled('electron-winstaller') &&
            !process.env.DISABLE_SQUIRREL_TEST);
    }
    async make({ dir, makeDir, targetArch, packageJSON, appName, forgeConfig, }) {
        const outPath = node_path_1.default.resolve(makeDir, `squirrel.windows/${targetArch}`);
        await this.ensureDirectory(outPath);
        const winstallerConfig = {
            name: typeof packageJSON.name === 'string'
                ? packageJSON.name.replace(/-/g, '_')
                : undefined, // squirrel hates hyphens
            title: appName,
            noMsi: true,
            exe: `${forgeConfig.packagerConfig.executableName || appName}.exe`,
            setupExe: `${appName}-${packageJSON.version} Setup.exe`,
            ...this.config,
            appDirectory: dir,
            outputDirectory: outPath,
        };
        await (0, electron_winstaller_1.createWindowsInstaller)(winstallerConfig);
        const nupkgVersion = (0, electron_winstaller_1.convertVersion)(packageJSON.version);
        const artifacts = [
            node_path_1.default.resolve(outPath, 'RELEASES'),
            node_path_1.default.resolve(outPath, winstallerConfig.setupExe || `${appName}Setup.exe`),
            node_path_1.default.resolve(outPath, `${winstallerConfig.name}-${nupkgVersion}-full.nupkg`),
        ];
        const deltaPath = node_path_1.default.resolve(outPath, `${winstallerConfig.name}-${nupkgVersion}-delta.nupkg`);
        if (winstallerConfig.remoteReleases &&
            !winstallerConfig.noDelta &&
            (await fs_extra_1.default.pathExists(deltaPath))) {
            artifacts.push(deltaPath);
        }
        const msiPath = node_path_1.default.resolve(outPath, winstallerConfig.setupMsi || `${appName}Setup.msi`);
        if (!winstallerConfig.noMsi && (await fs_extra_1.default.pathExists(msiPath))) {
            artifacts.push(msiPath);
        }
        return artifacts;
    }
}
exports.default = MakerSquirrel;
exports.MakerSquirrel = MakerSquirrel;
