// @electron-forge/plugin-vite/dist/ViteConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
const vite_1 = require("vite");
const vite_main_config_1 = require("./config/vite.main.config");
const vite_preload_config_1 = require("./config/vite.preload.config");
const vite_renderer_config_1 = require("./config/vite.renderer.config");
const d = (0, debug_1.default)('@electron-forge/plugin-vite:ViteConfig');
class ViteConfigGenerator {
    constructor(pluginConfig, projectDir, isProd) {
        this.pluginConfig = pluginConfig;
        this.projectDir = projectDir;
        this.isProd = isProd;
        d('Config mode:', this.mode);
    }
    async resolveConfig(buildConfig, target) {
        const configEnv = {
            // @see - https://vitejs.dev/config/#conditional-config
            command: this.isProd ? 'build' : 'serve',
            // `mode` affects `.env.[mode]` file load.
            mode: this.mode,
            // Forge extension variables.
            root: this.projectDir,
            forgeConfig: this.pluginConfig,
            forgeConfigSelf: buildConfig,
        };
        // `configEnv` is to be passed as an arguments when the user export a function in `vite.config.js`.
        const userConfig = (await (0, vite_1.loadConfigFromFile)(configEnv, buildConfig.config))
            ?.config;
        switch (target) {
            case 'main':
                return (0, vite_main_config_1.getConfig)(configEnv, userConfig);
            case 'preload':
                return (0, vite_preload_config_1.getConfig)(configEnv, userConfig);
            case 'renderer':
                return (0, vite_renderer_config_1.getConfig)(configEnv, userConfig);
            default:
                throw new Error(`Unknown target: ${target}, expected 'main', 'preload' or 'renderer'`);
        }
    }
    get mode() {
        // Vite's `mode` can be passed in via command.
        // Since we are currently using the JavaScript API, we are opinionated defining two default values for mode here.
        // The `mode` set by the end user in `vite.config.js` has a higher priority.
        return this.isProd ? 'production' : 'development';
    }
    async getBuildConfigs() {
        if (!Array.isArray(this.pluginConfig.build)) {
            throw new Error('"config.build" must be an Array');
        }
        const configs = this.pluginConfig.build
            // Prevent load the default `vite.config.js` file.
            .filter(({ config }) => config)
            .map((buildConfig) => this.resolveConfig(buildConfig, buildConfig.target ?? 'main'));
        return await Promise.all(configs);
    }
    async getRendererConfig() {
        if (!Array.isArray(this.pluginConfig.renderer)) {
            throw new Error('"config.renderer" must be an Array');
        }
        const configs = this.pluginConfig.renderer
            // Prevent load the default `vite.config.js` file.
            .filter(({ config }) => config)
            .map((buildConfig) => this.resolveConfig(buildConfig, 'renderer'));
        return await Promise.all(configs);
    }
}
exports.default = ViteConfigGenerator;
