// @electron-forge/template-webpack-typescript/dist/WebpackTypeScriptTemplate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const template_base_1 = require("@electron-forge/template-base");
const fs_extra_1 = __importDefault(require("fs-extra"));
class WebpackTypeScriptTemplate extends template_base_1.BaseTemplate {
    constructor() {
        super(...arguments);
        this.templateDir = node_path_1.default.resolve(__dirname, '..', 'tmpl');
    }
    async initializeTemplate(directory, options) {
        const superTasks = await super.initializeTemplate(directory, options);
        return [
            ...superTasks,
            {
                title: 'Setting up Forge configuration',
                task: async () => {
                    await this.copyTemplateFile(directory, 'forge.config.ts');
                    await fs_extra_1.default.remove(node_path_1.default.resolve(directory, 'forge.config.js'));
                },
            },
            {
                title: 'Preparing TypeScript files and configuration',
                task: async () => {
                    const filePath = (fileName) => node_path_1.default.join(directory, 'src', fileName);
                    // Copy Webpack files
                    await this.copyTemplateFile(directory, 'webpack.main.config.ts');
                    await this.copyTemplateFile(directory, 'webpack.renderer.config.ts');
                    await this.copyTemplateFile(directory, 'webpack.rules.ts');
                    await this.copyTemplateFile(directory, 'webpack.plugins.ts');
                    await this.updateFileByLine(node_path_1.default.resolve(directory, 'src', 'index.html'), (line) => {
                        if (line.includes('link rel="stylesheet"'))
                            return '';
                        return line;
                    });
                    // Copy tsconfig with a small set of presets
                    await this.copyTemplateFile(directory, 'tsconfig.json');
                    // Copy eslint config with recommended settings
                    await this.copyTemplateFile(directory, '.eslintrc.json');
                    // Remove index.js and replace with index.ts
                    await fs_extra_1.default.remove(filePath('index.js'));
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'index.ts');
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'renderer.ts');
                    // Remove preload.js and replace with preload.ts
                    await fs_extra_1.default.remove(filePath('preload.js'));
                    await this.copyTemplateFile(node_path_1.default.join(directory, 'src'), 'preload.ts');
                    // update package.json
                    const packageJSONPath = node_path_1.default.resolve(directory, 'package.json');
                    const packageJSON = await fs_extra_1.default.readJson(packageJSONPath);
                    packageJSON.main = '.webpack/main';
                    // Configure scripts for TS template
                    packageJSON.scripts.lint = 'eslint --ext .ts,.tsx .';
                    await fs_extra_1.default.writeJson(packageJSONPath, packageJSON, {
                        spaces: 2,
                    });
                },
            },
        ];
    }
}
exports.default = new WebpackTypeScriptTemplate();
