// @electron-forge/core/dist/util/hook.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runMutatingHook = exports.getHookListrTasks = exports.runHook = void 0;
const chalk_1 = __importDefault(require("chalk"));
const debug_1 = __importDefault(require("debug"));
const d = (0, debug_1.default)('electron-forge:hook');
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const runHook = async (forgeConfig, hookName, ...hookArgs) => {
    const { hooks } = forgeConfig;
    if (hooks) {
        d(`hook triggered: ${hookName}`);
        if (typeof hooks[hookName] === 'function') {
            d('calling hook:', hookName, 'with args:', hookArgs);
            await hooks[hookName](forgeConfig, ...hookArgs);
        }
    }
    await forgeConfig.pluginInterface.triggerHook(hookName, hookArgs);
};
exports.runHook = runHook;
const getHookListrTasks = async (childTrace, forgeConfig, hookName, ...hookArgs) => {
    const { hooks } = forgeConfig;
    const tasks = [];
    if (hooks) {
        d(`hook triggered: ${hookName}`);
        if (typeof hooks[hookName] === 'function') {
            d('calling hook:', hookName, 'with args:', hookArgs);
            tasks.push({
                title: `Running ${chalk_1.default.yellow(hookName)} hook from forgeConfig`,
                task: childTrace({
                    name: 'forge-config-hook',
                    category: '@electron-forge/hooks',
                    extraDetails: { hook: hookName },
                }, async () => {
                    await hooks[hookName](forgeConfig, ...hookArgs);
                }),
            });
        }
    }
    tasks.push(...(await forgeConfig.pluginInterface.getHookListrTasks(childTrace, hookName, hookArgs)));
    return tasks;
};
exports.getHookListrTasks = getHookListrTasks;
async function runMutatingHook(forgeConfig, hookName, ...item) {
    const { hooks } = forgeConfig;
    if (hooks) {
        d(`hook triggered: ${hookName}`);
        if (typeof hooks[hookName] === 'function') {
            d('calling mutating hook:', hookName, 'with item:', item[0]);
            const hook = hooks[hookName];
            const result = await hook(forgeConfig, ...item);
            if (typeof result !== 'undefined') {
                item[0] = result;
            }
        }
    }
    return forgeConfig.pluginInterface.triggerMutatingHook(hookName, item[0]);
}
exports.runMutatingHook = runMutatingHook;
