// @electron-forge/core/dist/util/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_utils_1 = require("@electron-forge/core-utils");
const forge_config_1 = require("./forge-config");
class ForgeUtils {
    constructor() {
        this.getElectronVersion = core_utils_1.getElectronVersion;
        this.spawnPackageManager = core_utils_1.spawnPackageManager;
    }
    /**
     * Helper for creating a dynamic config value that will get its real value
     * based on the "buildIdentifier" in your Forge config.
     *
     * Usage:
     * `fromBuildIdentifier({ stable: 'App', beta: 'App Beta' })`
     */
    fromBuildIdentifier(map) {
        return (0, forge_config_1.fromBuildIdentifier)(map);
    }
    /**
     * Register a virtual config file for forge to find.
     * Takes precedence over other configuration options like a forge.config.js file.
     * Dir should point to the folder containing the app.
     */
    registerForgeConfigForDirectory(dir, config) {
        return (0, forge_config_1.registerForgeConfigForDirectory)(dir, config);
    }
    /**
     * Unregister a forge config previously registered with registerForgeConfigForDirectory.
     */
    unregisterForgeConfigForDirectory(dir) {
        return (0, forge_config_1.unregisterForgeConfigForDirectory)(dir);
    }
}
exports.default = ForgeUtils;
