// @electron-forge/core/dist/util/install-dependencies.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DepVersionRestriction = exports.DepType = void 0;
const core_utils_1 = require("@electron-forge/core-utils");
const cross_spawn_promise_1 = require("@malept/cross-spawn-promise");
const debug_1 = __importDefault(require("debug"));
const d = (0, debug_1.default)('electron-forge:dependency-installer');
var DepType;
(function (DepType) {
    DepType["PROD"] = "PROD";
    DepType["DEV"] = "DEV";
})(DepType || (exports.DepType = DepType = {}));
var DepVersionRestriction;
(function (DepVersionRestriction) {
    DepVersionRestriction["EXACT"] = "EXACT";
    DepVersionRestriction["RANGE"] = "RANGE";
})(DepVersionRestriction || (exports.DepVersionRestriction = DepVersionRestriction = {}));
exports.default = async (pm, dir, deps, depType = DepType.PROD, versionRestriction = DepVersionRestriction.RANGE) => {
    d('installing', JSON.stringify(deps), 'in:', dir, `depType=${depType},versionRestriction=${versionRestriction},withPackageManager=${pm.executable}`);
    if (deps.length === 0) {
        d('nothing to install, stopping immediately');
        return Promise.resolve();
    }
    const cmd = [pm.install].concat(deps);
    if (depType === DepType.DEV)
        cmd.push(pm.dev);
    if (versionRestriction === DepVersionRestriction.EXACT)
        cmd.push(pm.exact);
    d('executing', JSON.stringify(cmd), 'in:', dir);
    try {
        await (0, core_utils_1.spawnPackageManager)(pm, cmd, {
            cwd: dir,
            stdio: 'pipe',
        });
    }
    catch (err) {
        if (err instanceof cross_spawn_promise_1.ExitError) {
            throw new Error(`Failed to install modules: ${JSON.stringify(deps)}\n\nWith output: ${err.message}\n${err.stderr ? err.stderr.toString() : ''}`);
        }
        else {
            throw err;
        }
    }
};
