// @electron-forge/core/dist/util/publish-state.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_crypto_1 = __importDefault(require("node:crypto"));
const node_path_1 = __importDefault(require("node:path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const EXTENSION = '.forge.publish';
class PublishState {
    static async loadFromDirectory(directory, rootDir) {
        if (!(await fs_extra_1.default.pathExists(directory))) {
            throw new Error(`Attempted to load publish state from a missing directory: ${directory}`);
        }
        const publishes = [];
        for (const dirName of await fs_extra_1.default.readdir(directory)) {
            const subDir = node_path_1.default.resolve(directory, dirName);
            const states = [];
            if ((await fs_extra_1.default.stat(subDir)).isDirectory()) {
                const filePaths = (await fs_extra_1.default.readdir(subDir))
                    .filter((fileName) => fileName.endsWith(EXTENSION))
                    .map((fileName) => node_path_1.default.resolve(subDir, fileName));
                for (const filePath of filePaths) {
                    const state = new PublishState(filePath);
                    await state.load();
                    state.state.artifacts = state.state.artifacts.map((artifactPath) => node_path_1.default.resolve(rootDir, artifactPath));
                    states.push(state);
                }
            }
            publishes.push(states);
        }
        return publishes;
    }
    static async saveToDirectory(directory, artifacts, rootDir) {
        const id = node_crypto_1.default
            .createHash('SHA256')
            .update(JSON.stringify(artifacts))
            .digest('hex');
        for (const artifact of artifacts) {
            artifact.artifacts = artifact.artifacts.map((artifactPath) => node_path_1.default.relative(rootDir, artifactPath));
            const publishState = new PublishState(node_path_1.default.resolve(directory, id, 'null'), false);
            publishState.state = artifact;
            await publishState.saveToDisk();
        }
    }
    constructor(filePath, hasHash = true) {
        this.state = {};
        this.dir = node_path_1.default.dirname(filePath);
        this.path = filePath;
        this.hasHash = hasHash;
    }
    generateHash() {
        const content = JSON.stringify(this.state || {});
        return node_crypto_1.default.createHash('SHA256').update(content).digest('hex');
    }
    async load() {
        this.state = await fs_extra_1.default.readJson(this.path);
    }
    async saveToDisk() {
        if (!this.hasHash) {
            this.path = node_path_1.default.resolve(this.dir, `${this.generateHash()}${EXTENSION}`);
            this.hasHash = true;
        }
        await fs_extra_1.default.mkdirs(node_path_1.default.dirname(this.path));
        await fs_extra_1.default.writeJson(this.path, this.state);
    }
}
exports.default = PublishState;
