// @electron-forge/maker-rpm/dist/MakerRpm.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakerRpm = exports.rpmArch = void 0;
const node_path_1 = __importDefault(require("node:path"));
const maker_base_1 = require("@electron-forge/maker-base");
function renameRpm(dest, _src) {
    return node_path_1.default.join(dest, '<%= name %>-<%= version %>-<%= revision %>.<%= arch === "aarch64" ? "arm64" : arch %>.rpm');
}
function rpmArch(nodeArch) {
    switch (nodeArch) {
        case 'ia32':
            return 'i386';
        case 'x64':
            return 'x86_64';
        case 'arm64':
            return 'aarch64';
        case 'armv7l':
            return 'armv7hl';
        case 'arm':
            return 'armv6hl';
        default:
            return nodeArch;
    }
}
exports.rpmArch = rpmArch;
class MakerRpm extends maker_base_1.MakerBase {
    constructor() {
        super(...arguments);
        this.name = 'rpm';
        this.defaultPlatforms = ['linux'];
        this.requiredExternalBinaries = ['rpmbuild'];
    }
    isSupportedOnCurrentPlatform() {
        return this.isInstalled('electron-installer-redhat');
    }
    async make({ dir, makeDir, targetArch }) {
        // eslint-disable-next-line n/no-missing-require
        const installer = require('electron-installer-redhat');
        const outDir = node_path_1.default.resolve(makeDir, 'rpm', targetArch);
        await this.ensureDirectory(outDir);
        const { packagePaths } = await installer({
            ...this.config,
            arch: rpmArch(targetArch),
            src: dir,
            dest: outDir,
            rename: renameRpm,
        });
        return packagePaths;
    }
}
exports.default = MakerRpm;
exports.MakerRpm = MakerRpm;
