// @electron-forge/plugin-fuses/dist/FusesPlugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FusesPlugin = void 0;
const node_path_1 = __importDefault(require("node:path"));
const fuses_1 = require("@electron/fuses");
const plugin_base_1 = require("@electron-forge/plugin-base");
const getElectronExecutablePath_1 = require("./util/getElectronExecutablePath");
class FusesPlugin extends plugin_base_1.PluginBase {
    constructor(fusesConfig) {
        super(fusesConfig);
        this.name = 'fuses';
        this.fusesConfig = {};
        this.fusesConfig = fusesConfig;
    }
    getHooks() {
        return {
            packageAfterCopy: (0, plugin_base_1.namedHookWithTaskFn)(async (listrTask, resolvedForgeConfig, resourcesPath, electronVersion, platform, arch) => {
                const { fusesConfig } = this;
                const applePlatforms = ['darwin', 'mas'];
                if (Object.keys(fusesConfig).length) {
                    const pathToElectronExecutable = (0, getElectronExecutablePath_1.getElectronExecutablePath)({
                        appName: applePlatforms.includes(platform)
                            ? 'Electron'
                            : 'electron',
                        basePath: node_path_1.default.resolve(resourcesPath, '../..'),
                        platform,
                    });
                    const osxSignConfig = resolvedForgeConfig.packagerConfig.osxSign;
                    const hasOSXSignConfig = (typeof osxSignConfig === 'object' &&
                        Boolean(Object.keys(osxSignConfig).length)) ||
                        Boolean(osxSignConfig);
                    await (0, fuses_1.flipFuses)(pathToElectronExecutable, {
                        resetAdHocDarwinSignature: !hasOSXSignConfig &&
                            applePlatforms.includes(platform) &&
                            arch === 'arm64',
                        ...this.fusesConfig,
                    });
                }
            }, 'Flipping Fuses'),
        };
    }
}
exports.default = FusesPlugin;
exports.FusesPlugin = FusesPlugin;
