// @electron-forge/plugin-base/dist/Plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginBase = exports.namedHookWithTaskFn = void 0;
class Plugin {
    constructor(config) {
        this.config = config;
        /** @internal */
        this._resolvedHooks = {};
        Object.defineProperty(this, '__isElectronForgePlugin', {
            value: true,
            enumerable: false,
            configurable: false,
        });
    }
    init(_dir, _config) {
        // This logic ensures that we only call getHooks once regardless of how many
        // times we trip hook logic in the PluginInterface.
        this._resolvedHooks = this.getHooks();
        this.getHooks = () => this._resolvedHooks;
    }
    getHooks() {
        return {};
    }
    async startLogic(_startOpts) {
        return false;
    }
}
exports.default = Plugin;
exports.PluginBase = Plugin;
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 *
 * This is a filthy hack around TypeScript to allow internal hooks in our
 * internal plugins to have some level of access to the "Task" that Listr2 runs.
 * Specifically the ability to set a custom task name and receive the task
 *
 * This method is not type-safe internally, but is type-safe for consumers.
 *
 * @internal
 */
const namedHookWithTaskFn = (hookFn, name) => {
    function namedHookWithTaskInner(...args) {
        return hookFn(this, ...args);
    }
    const fn = namedHookWithTaskInner;
    fn.__hookName = name;
    return fn;
};
exports.namedHookWithTaskFn = namedHookWithTaskFn;
