// @electron-forge/plugin-vite/dist/config/vite.preload.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = void 0;
const vite_1 = require("vite");
const vite_base_config_1 = require("./vite.base.config");
function getConfig(forgeEnv, userConfig = {}) {
    const { forgeConfigSelf } = forgeEnv;
    const config = {
        build: {
            copyPublicDir: false,
            rollupOptions: {
                external: [...vite_base_config_1.external, 'electron/renderer'],
                // Preload scripts may contain Web assets, so use the `build.rollupOptions.input` instead `build.lib.entry`.
                input: forgeConfigSelf.entry,
                output: {
                    format: 'cjs',
                    // It should not be split chunks.
                    inlineDynamicImports: true,
                    entryFileNames: '[name].js',
                    chunkFileNames: '[name].js',
                    assetFileNames: '[name].[ext]',
                },
            },
        },
        plugins: [(0, vite_base_config_1.pluginHotRestart)('reload')],
    };
    const buildConfig = (0, vite_base_config_1.getBuildConfig)(forgeEnv);
    return (0, vite_1.mergeConfig)((0, vite_1.mergeConfig)(buildConfig, config), userConfig);
}
exports.getConfig = getConfig;
