// @electron-forge/plugin-vite/dist/config/vite.base.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginHotRestart = exports.pluginExposeRenderer = exports.getBuildDefine = exports.getDefineKeys = exports.getBuildConfig = exports.external = exports.builtins = void 0;
const node_module_1 = require("node:module");
exports.builtins = [
    'electron',
    'electron/common',
    ...node_module_1.builtinModules.map((m) => [m, `node:${m}`]).flat(),
];
exports.external = [...exports.builtins];
// Used for hot reload after preload scripts.
const viteDevServers = {};
const viteDevServerUrls = {};
function getBuildConfig(env) {
    const { root, mode, command } = env;
    return {
        root,
        mode,
        build: {
            // Prevent multiple builds from interfering with each other.
            emptyOutDir: false,
            // 🚧 Multiple builds may conflict.
            outDir: '.vite/build',
            watch: command === 'serve' ? {} : null,
            minify: command === 'build',
        },
        clearScreen: false,
    };
}
exports.getBuildConfig = getBuildConfig;
function getDefineKeys(names) {
    const define = {};
    // change name from kebab case to upper snake case to agree with vite:define plugin
    // this allows the VitePluginRendererConfig entries to contain names with dashes
    return names.reduce((acc, name) => {
        const NAME = name.toUpperCase().replaceAll('-', '_');
        const keys = {
            VITE_DEV_SERVER_URL: `${NAME}_VITE_DEV_SERVER_URL`,
            VITE_NAME: `${NAME}_VITE_NAME`,
        };
        return { ...acc, [name]: keys };
    }, define);
}
exports.getDefineKeys = getDefineKeys;
function getBuildDefine(env) {
    const { command, forgeConfig } = env;
    const names = forgeConfig.renderer
        .filter(({ name }) => name != null)
        .map(({ name }) => name);
    const defineKeys = getDefineKeys(names);
    const define = Object.entries(defineKeys).reduce((acc, [name, keys]) => {
        const { VITE_DEV_SERVER_URL, VITE_NAME } = keys;
        const def = {
            [VITE_DEV_SERVER_URL]: command === 'serve'
                ? JSON.stringify(viteDevServerUrls[VITE_DEV_SERVER_URL])
                : undefined,
            [VITE_NAME]: JSON.stringify(name),
        };
        return { ...acc, ...def };
    }, {});
    return define;
}
exports.getBuildDefine = getBuildDefine;
function pluginExposeRenderer(name) {
    const { VITE_DEV_SERVER_URL } = getDefineKeys([name])[name];
    return {
        name: '@electron-forge/plugin-vite:expose-renderer',
        configureServer(server) {
            // Expose server for preload scripts hot reload.
            viteDevServers[name] = server;
            server.httpServer?.once('listening', () => {
                const addressInfo = server.httpServer?.address();
                // Expose env constant for main process use.
                viteDevServerUrls[VITE_DEV_SERVER_URL] =
                    `http://localhost:${addressInfo?.port}`;
            });
        },
    };
}
exports.pluginExposeRenderer = pluginExposeRenderer;
function pluginHotRestart(command) {
    return {
        name: '@electron-forge/plugin-vite:hot-restart',
        closeBundle() {
            if (command === 'reload') {
                for (const server of Object.values(viteDevServers)) {
                    // Preload scripts hot reload.
                    server.ws.send({ type: 'full-reload' });
                }
            }
            else if (command === 'restart') {
                // Main process hot restart.
                // https://github.com/electron/forge/blob/v7.2.0/packages/api/core/src/api/start.ts#L216-L223
                // TODO: blocked in #3380
                // process.stdin.emit('data', 'rs');
            }
        },
    };
}
exports.pluginHotRestart = pluginHotRestart;
