// @electron-forge/plugin-vite/dist/config/vite.main.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = void 0;
const vite_1 = require("vite");
const vite_base_config_1 = require("./vite.base.config");
function getConfig(forgeEnv, userConfig = {}) {
    const { forgeConfigSelf } = forgeEnv;
    const define = (0, vite_base_config_1.getBuildDefine)(forgeEnv);
    const config = {
        build: {
            copyPublicDir: false,
            rollupOptions: {
                external: [...vite_base_config_1.external, 'electron/main'],
            },
        },
        plugins: [(0, vite_base_config_1.pluginHotRestart)('restart')],
        define,
        resolve: {
            // Load the Node.js entry.
            conditions: ['node'],
            mainFields: ['module', 'jsnext:main', 'jsnext'],
        },
    };
    const buildConfig = (0, vite_base_config_1.getBuildConfig)(forgeEnv);
    if (userConfig.build?.lib == null) {
        config.build.lib = {
            entry: forgeConfigSelf.entry,
            fileName: () => '[name].js',
            formats: ['cjs'],
        };
    }
    return (0, vite_1.mergeConfig)((0, vite_1.mergeConfig)(buildConfig, config), userConfig);
}
exports.getConfig = getConfig;
